Filter users displayed on Project Performance Dashboard by groups selected in filter panel

rickypchen opened the pull request:
@esoergel @czue 
From feedback google doc for [PPD](https://docs.google.com/document/d/1lpGuOcMI7sXUkKxdUVz_-Vbwy77vVjrENvoSF5Axsnw/edit#)

kaapstorm committed:
Use "select1" data type for choices

kaapstorm committed:
XFormBuilder supports hidden values

kaapstorm committed:
XFormBuilder supports default value

kaapstorm committed:
Fix case list

kaapstorm committed:
Add event forms to subject module

kaapstorm committed:
Use study name as comment, not app name

kaapstorm committed:
Add study description and unique protocol ID

kaapstorm committed:
Improved case management

Conflicts:
	custom/openclinica/management/commands/odm_to_app.py

kaapstorm committed:
Make all form unique IDs unique

kaapstorm committed:
Use constants for case properties

kaapstorm committed:
XFormBuilder support for more question params

kaapstorm committed:
Add range check validation support

kaapstorm committed:
Fixed form unique ID

kaapstorm committed:
Updated tests

kaapstorm committed:
Documentation updated

kaapstorm committed:
Fixes and tests

kaapstorm committed:
Missed a bunch of constants

kaapstorm committed:
Deleted old test file

kaapstorm committed:
Remove xpath duplication. Add doctest

kaapstorm committed:
Move Register Subject to its own module

kaapstorm committed:
Generate question hints from ODM comments

Commit:
add property use_j2me_endpoint to ApplicationBase

Commit:
set use_j2me_endpoint in app settings UI

Commit:
use use_j2me_endpoint to determine whether to replace the endpoint for j2me builds

rickypchen committed:
added group filter panel to project_health_dashboard

czue committed:
extract function to analytics module

czue committed:
return a dict instead of a list of tuples

czue committed:
add basic tests for get_ledger_values_for_case_as_of

rickypchen committed:
added group filter functionality

rickypchen committed:
removed testing html display

rickypchen committed:
removed console.log for testing

rickypchen committed:
changed number_of_performing_users to property and fixed delta_performing_pct display when 0

dimagimon commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,50 +47,79 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
     @property
     def delta_performing(self):
-        return self.performing - self._previous_summary.performing if self._previous_summary else self.performing
+        return self.number_of_performing_users - self._previous_summary.number_of_performing_users if self._previous_summary else self.number_of_performing_users
```
E501 line too long (161 > 115 characters)

dimagimon commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,50 +47,79 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
+
+    @property
+    def has_group_filter(self):
+        if self._users_filtered_by_group: return True; return False
```
E701 multiple statements on one line (colon)
E702 multiple statements on one line (semicolon)

dimagimon commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -106,12 +134,29 @@ def get_all_user_stubs(self):
-            user_stubs = self.get_all_user_stubs()
+            if self.has_group_filter:
+                previous_stubs = self._previous_summary.get_all_user_stubs_by_filtered_group()
+                next_stubs = self._next_summary.get_all_user_stubs_by_filtered_group() if self._next_summary else {}
```
E501 line too long (116 > 115 characters)

rickypchen committed:
fixed delta_performing_pct divide by zero

dimagimon commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -106,12 +139,29 @@ def get_all_user_stubs(self):
-            user_stubs = self.get_all_user_stubs()
+            if self._users_filtered_by_group:
+                previous_stubs = self._previous_summary.get_all_user_stubs_by_filtered_group()
+                next_stubs = self._next_summary.get_all_user_stubs_by_filtered_group() if self._next_summary else {}
```
E501 line too long (116 > 115 characters)

rickypchen committed:
fixed pylint errors

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,49 +49,81 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
         )
+        self._performing_queryset_group_filtered = self._apply_group_filter.filter(
+            num_of_forms__gte=performance_threshold,
+        )
```
can you clarify why you need both of these? can you not just apply the user filter if specified and otherwise not?

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,49 +49,81 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
     def delta_performing_pct(self):
-        if self.delta_performing and self._previous_summary and self._previous_summary.performing:
-            return float(self.delta_performing / float(self._previous_summary.performing)) * 100.
+        if self._previous_summary.number_of_performing_users:
```
do you not still have to have to check for `None` here?

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,49 +49,81 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
+        if self._previous_summary.number_of_performing_users:
+            return float(self.delta_performing / float(self._previous_summary.number_of_performing_users)) * 100.
+        else:
+            return 100
```
maybe related to the above - this makes sense if `self._previous_summary.number_of_performing_users` is equal to `0`, but if `self._previous_summary` isn't set at all then returning `100` seems a bit misleading.

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,49 +49,81 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
         )
+        self._apply_group_filter = self._base_queryset.filter(
+            user_id__in=users_filtered_by_group,
+        )
```
same comment as below - could this not just be applied to `base_queryset`?

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -184,12 +241,28 @@ def show_in_navigation(cls, domain=None, project=None, user=None):
+
+    def get_users_by_filtered_groupids(self):
+        groupids_param = self.get_filtered_group_ids()
+        users_list = GroupES().domain('ccdt').group_ids(groupids_param).fields(["users"]).values()
```
this should be `self.domain`

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -184,12 +241,28 @@ def show_in_navigation(cls, domain=None, project=None, user=None):
+        user_id_list = []
+        for user in users_list:
+            usersid = user.values()[0]
+            if type(usersid) is list:
```
when do you expect a `list` versus a `string`?

nit: `isinstance(usersid, list)` would likely be better

rickypchen replied:
If the group has more than one user, then I would expect usersid to be a list, and a string if the group only has one user. I definitely should look out for inheritance, so thank you for pointing out `isinstance` function!

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,49 +49,81 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
     @property
     def delta_active_pct(self):
-        if self.delta_active and self._previous_summary and self._previous_summary.active:
+        if self._previous_summary.active:
```
(same question about `self._previous_summary` being `None`)

czue commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -198,7 +271,10 @@ def template_context(self):
+                users_filtered_by_group=users_in_group,
             )
+            this_month_summary.set_performing(this_month_summary.number_of_performing_users)
+            this_month_summary.set_active_users(this_month_summary.number_of_active_users)
```
why did these need to move outside the constructor?

czue commented on corehq/apps/reports/templates/reports/project_health/project_health_dashboard.html:
```diff
@@ -13,10 +13,18 @@
+    <script src="{% new_static 'reports/javascripts/reports.config.js' %}"></script>
+    <script src="{% new_static 'reports/javascripts/reports.util.js' %}"></script>
+    <script src="{% new_static 'reports/javascripts/reports.async.js' %}"></script>
+{% endblock %}
```
didn't look closely at this, but is there any chance it's easy to have this extend one of the core reporting base templates? seems like we could find ourselves in a situation where it forks a lot of the same code if it's reimplementing a lot of that stuff (this is 100% my fault for starting it outside the framework, but something I was hoping might change with this work)

czue commented:
thanks @rickypchen! i had a bunch of pretty minor comments but this looks like it's coming along nicely

kaapstorm committed:
OpenClinica SOAP API client

Commit:
put widget behind toggle

rickypchen committed:
refactoring and removing self._performing_queryset_group_filtered

Commit:
remove unused imports from management commands

millerdev committed:
Do not use read-only commcare-hq mount on travis

proteusvacuum committed:
only add stock response for relevant cases

esoergel committed:
Allow target=save in media_profile.ccpr dl

esoergel commented on corehq/apps/reports/standard/project_health.py:
```diff
@@ -48,49 +49,81 @@ class MonthlyPerformanceSummary(jsonobject.JsonObject):
-    def __init__(self, domain, month, performance_threshold, previous_summary=None):
+    def __init__(self, domain, month, users_filtered_by_group, performance_threshold, previous_summary=None):
         self._previous_summary = previous_summary
+        self._users_filtered_by_group = users_filtered_by_group
```
Similar to what Cory suggested - you could only fork the logic here in the constructor, then the changes applied to `_base_queryset` will be available everywhere necessary.  That is:
```python
if users_filtered_by_group:
    # add additional filter to _base_queryset
```
Then you should be able to just always refer to `_base_queryset` without needing to check `users_filtered_by_group`

rickypchen committed:
refactoring out _apply_group_filter so that filter applies to base_queryset

rickypchen committed:
refactoring out _apply_group_filter so that filter applies to base_queryset

esoergel commented on corehq/apps/reports/templates/reports/project_health/project_health_dashboard.html:
```diff
@@ -103,7 +115,7 @@
             <p class="text-muted"><small>{% blocktrans %}Users who submitted {{ threshold }} or more forms last month.{% endblocktrans %}</small></p>
         </td>
-        <td>{{ last_month.performing }}</td>
+        <td>{{ last_month.number_of_performing_users }}</td>
```
:+1: on the name change

rickypchen committed:
forked filter logic into constructor

rickypchen committed:
forked filter logic into constructor

proteusvacuum committed:
This is not a test

snopoke committed:
exlude rows with no doc

Commit:
Merge pull request #12019 from dimagi/reindex

exlude rows with no doc

Commit:
Add reindex commands for sql cases and forms

snopoke committed:
Merge pull request #12020 from dimagi/reindex-sql-forms-cases

Add reindex commands for sql cases and forms

millerdev committed:
docker-compose 1.5.2 for variable substitution

Commit:
fix a bunch of tests in Django 1.8

Commit:
use clearer text when feature usage is over limit

czue committed:
Merge pull request #12005 from dimagi/clear-domain-has-apps-cache

Use _rev instead of _id when checking to clear cache

czue committed:
Merge pull request #11989 from dimagi/fix-time-since

accept 0 as being valid field value

Commit:
Added facets to commcare version report

czue committed:
ensure correct domain used

snopoke committed:
rebuild schema's on error

czue committed:
fix bad variable ref

czue committed:
add entries function

snopoke committed:
exclude system forms from counts by user

snopoke committed:
rename imports

snopoke committed:
exclude usercases from report case counts

czue committed:
make datetime a property

snopoke committed:
filter by case type if supplied

snopoke committed:
exclude user case

czue committed:
[ci skip] lint

snopoke committed:
add error message

snopoke committed:
log exception if exception

snopoke committed:
don't die if no schema

snopoke committed:
Merge pull request #12023 from dimagi/loc-inventory-to-es

minor cleanup and tests to inventory by location report

snopoke committed:
Merge pull request #12004 from dimagi/trap-form-schema-error

Trap errors encountered during get_form_data_schema

czue committed:
fix reporting rate report crashes

http://manage.dimagi.com/default.asp?229623

czue committed:
remove date filter from aggregate inventory report

czue committed:
remove date filter from CurrentStockStatusReport

czue committed:
clean up comments

czue committed:
remove date filter from stock status map report

czue committed:
remove commented out line

czue committed:
cleanup and fix tests

snopoke committed:
exclude deleted cases part 1

snopoke committed:
exclude deleted cases part 2

snopoke committed:
use extracted method

snopoke committed:
exclude deleted cases part 3

snopoke committed:
exclude deleted cases part 4

czue committed:
Merge pull request #12022 from dimagi/rebuild-schema

rebuild schema's on error

snopoke committed:
Merge pull request #12025 from dimagi/reporting-rate-crash

fix reporting rate report crashes

czue committed:
Merge pull request #12026 from dimagi/kill-date-filters

Kill date filters on supply reports

Commit:
Merge branch 'master' into 133676-case

Conflicts:
	corehq/toggles.py

Commit:
Merge branch 'master' into 228611-case

NoahCarnahan committed:
submodule update

NoahCarnahan committed:
Merge remote-tracking branch 'origin/master' into prelogin-link

snopoke committed:
update deploy

snopoke committed:
Merge pull request #12031 from dimagi/prelogin-link

Change sign-up link on pricing page to registration page

sravfeyn committed:
rewrite boolean logic

rickypchen committed:
added additional filter to _base_queryset if users_filtered_by_group

rickypchen committed:
added additional filter to _base_queryset if users_filtered_by_group

orangejenny committed:
Merge pull request #12024 from dimagi/system

exclude system forms and 'usercase' from reports

rickypchen committed:
reinstated number_of_performing_users property

rickypchen committed:
reinstated number_of_performing_users property

snopoke committed:
use_select2 in all EditUserViews

Commit:
Merge pull request #12017 from dimagi/target-save

Allow target=save in media_profile.ccpr dl

kaapstorm committed:
Updated tests

Commit:
remove testrunner.py

snopoke committed:
Merge pull request #12007 from dimagi/ucr-notify

add feature flag to get notified when a data source rebuild succeeds/fails

snopoke committed:
open in new window if requested

snopoke committed:
exclude data attribs when not app view

esoergel committed:
Merge pull request #12021 from dimagi/django-18-tests

fix a bunch of tests in Django 1.8

esoergel committed:
Merge pull request #12014 from dimagi/remove-unused-imports-custom

remove unused imports from custom/

emord committed:
Merge pull request #11965 from lwyszomi/hq_version_report

Added Commcare version report

Commit:
move filtering to all_cards

kaapstorm committed:
Encrypt password

kaapstorm committed:
Get study metadata from web service

dannyroberts committed:
fix delete_couch_devicelogs memory issue

- remove accidental include_docs=True when trying to get just ids

Commit:
support static UCR data in clone_domain

esoergel committed:
Merge pull request #12033 from dimagi/remove-testrunner.py

remove testrunner.py

dannyroberts committed:
Merge pull request #12037 from dimagi/delete_couch_devicelogs-fix

fix delete_couch_devicelogs memory issue

emord committed:
Merge pull request #12011 from dimagi/j2me-toggle

add widget to choose whether j2me should use http

Commit:
update staging.yaml

[ci skip]

Commit:
Update submodules

rickypchen committed:
added has_group_filter flag for when selected group has no users

rickypchen committed:
added has_group_filter flag for when selected group has no users

orangejenny committed:
Merge pull request #12034 from dimagi/open-new

allow open in new tab

orangejenny committed:
Merge pull request #12032 from dimagi/select2

use_select2 in all EditUserViews

Commit:
Update submodules

esoergel committed:
Merge pull request #12012 from dimagi/fix-nonetype

fix NoneType error

esoergel committed:
Merge pull request #12035 from dimagi/features-over-limit

use clearer text when feature usage is over limit

dannyroberts committed:
wrap <script> contents in onReady callback

esoergel committed:
Add command to force zone-aware replication

As described here https://www.elastic.co/guide/en/elasticsearch/reference/current/allocation-awareness.html

Commit:
Merge pull request #12038 from dimagi/static-ucr-clone_domain

support static UCR data in clone_domain

Commit:
Update name of report

Commit:
Fix name of report

Commit:
Merge branch 'master' into 133676-case

Conflicts:
	corehq/toggles.py

czue committed:
Merge pull request #12043 from dimagi/icds-static-ucr-patch

Update Names for ICDS Static Reports

snopoke committed:
Merge pull request #12041 from dimagi/allocation-awareness

Add command to force zone-aware replication

kaapstorm committed:
Hide API username and password from debugging

snopoke committed:
Merge pull request #11995 from dimagi/sync-payload-refactor

Sync payload refactor

snopoke committed:
Merge branch 'master' into publish-changes

snopoke committed:
add check option

snopoke committed:
test archive publishes change

snopoke committed:
test delete events published

snopoke committed:
test change rebuild event published

snopoke committed:
check deleted state

snopoke committed:
lint [ci skip]

NoahCarnahan committed:
Merge pull request #10507 from dimagi/openclinica

OpenClinica settings in project settings

snopoke committed:
set parent type correctly on location types

czue committed:
Merge pull request #11957 from dimagi/publish-changes

Publish changes

czue committed:
Merge pull request #12044 from dimagi/clone-domain

add 'check' option to clone domain

millerdev committed:
Add missing Kafka topics to topics.ALL

snopoke committed:
update langcodes

orangejenny committed:
Merge pull request #11874 from dimagi/biometric_case_list

Biometric search from case list

snopoke committed:
add required template variable

Commit:
Fix deploy link in firefox

snopoke committed:
Merge pull request #12047 from dimagi/fix-firefox-deploy-link

Fix deploy link in firefox

NoahCarnahan committed:
Handle document update conflict

millerdev committed:
Fix misc issues with docker setup

- Compile all pyc files for speed.
- Create empty staticfiles overlay with correct permissions so tests can write to it.
- Remove .log and .lock files so they can be recreated by test runner.
- Remove broken option used by docker teardown command.

snopoke committed:
use ugettext_lazy

snopoke committed:
Merge pull request #10516 from dimagi/odm_to_app

Improvements to generating apps from CDISC ODM docs

esoergel committed:
Merge pull request #12013 from dimagi/remove-unused-imports-management

remove unused imports from management commands

Commit:
change setUp to setUpClass and add super

millerdev committed:
Merge pull request #11743 from dimagi/session-api

Session api

Commit:
add more set up and tear down methods

esoergel committed:
Use better names for zone aware replication

[ci skip]

millerdev committed:
More misc issues

- Do not fail if log/lock files are missing.
- Compile custom and testapps (and hide output).
- Use pseudo-tty when running tests (fixes output issues with boot2docker).
- Write log file at path that can be written by cchq.

millerdev committed:
Add timeout to wait.sh in docker web container

millerdev committed:
Remove unused labels

orangejenny committed:
Update CASE_REFERENCE_VALIDATION_ERROR copy

snopoke committed:
simplify alembic diffs

millerdev committed:
Merge pull request #12051 from dimagi/es-cmd-names

Use better names for zone aware replication

snopoke committed:
Merge pull request #12053 from dimagi/orangejenny-patch-1

Update CASE_REFERENCE_VALIDATION_ERROR copy

Commit:
use setUp instead of setUpClass and tearDownClass

Commit:
update touchforms

Commit:
update dimagi utils

NoahCarnahan committed:
Merge pull request #12036 from dimagi/metadata_from_ws

Get study metadata using the OpenClinica web service

benrudolph committed:
Merge pull request #11942 from dimagi/upgrade-django-compressor

upgrade django-compressor

millerdev committed:
Fix compress test: do not open template for write

czue committed:
remove redundant overrides

czue committed:
delete commented code

Commit:
Update mpr_2bi_preg_delivery_death_list.json

Commit:
Update mpr_2bii_child_death_list.json

Commit:
Update mpr_2ci_child_birth_list.json

czue committed:
Revert "Session api"

Commit:
Merge pull request #12056 from dimagi/revert-11743-session-api

Revert "Session api"

orangejenny committed:
Merge pull request #12001 from dimagi/ds-not-found

handle DataSourceConfigurationNotFound

NoahCarnahan committed:
handle conflict for form schemas too

orangejenny committed:
Merge pull request #12027 from dimagi/exclude-delted

Exclude deleted cases and indices

czue committed:
delete ReportResource

Commit:
Merge pull request #12057 from dimagi/ct-cleanup

misc commtrack cleanup

millerdev committed:
Increase kafka topic creation timeout

Commit:
expose http endpoint checkbox to all users

millerdev committed:
Move coveralls out of docker (only used by travis)

esoergel committed:
Add pending_tasks command [ci skip]

Commit:
update staging.yaml

[ci skip]

calellowitz committed:
adding km tracking for create account button

NoahCarnahan committed:
Update translations

NoahCarnahan committed:
Merge branch 'master' of github.com:dimagi/commcare-hq

millerdev committed:
Merge pull request #12050 from dimagi/moar-setUp

fix more tests for django 1.8

millerdev committed:
Merge pull request #12054 from dimagi/update-utils

update dimagi utils

millerdev committed:
Merge pull request #12060 from dimagi/pending-tasks

Add pending_tasks command [ci skip]

NoahCarnahan committed:
Only save if needed

Commit:
Merge pull request #12055 from dimagi/icds-translations-for-ucr

ICDS Static UCR - Add Translations to List Reports

Commit:
Fix reference to wrong report ID and columns

esoergel committed:
Merge pull request #12059 from dimagi/remove-j2me-toggle

expose http endpoint checkbox to all users

Commit:
Wrap Stock Report in atomic transaction

czue committed:
Merge pull request #12048 from dimagi/export-conflict

Handle document update conflict when builder export schema

czue committed:
Merge pull request #12062 from dimagi/icds-ucr-fix

ICDS: Fix reference to wrong report ID and columns

snopoke committed:
Merge pull request #12058 from dimagi/kill-report-resource

delete ReportResource

czue committed:
remove commented code

czue committed:
pull out logic to get stock states into its own method

czue committed:
let leaf_node_data get its own list of stock states

czue committed:
use sql products

czue committed:
use sql_location

czue committed:
unused imports from commtrack models

Commit:
Move delete button to right div

czue committed:
extract function

czue committed:
add consumption helper and further extract function there

czue committed:
extract get_daily_consumption and get_monthly_consumption to helper

czue committed:
extract months_remaining to helper

czue committed:
add docstring, memoize possibly expensive function

czue committed:
extract resupply quantities

czue committed:
move stock_category to consumption helper

czue committed:
access consumption data through consumption helper object

czue committed:
[wip] add a layer of abstraction between StockStates and report data

this only affects one call-path through the data source but is working

czue committed:
unused import

czue committed:
move helper class to utils module

czue committed:
add doc string

czue committed:
move product_ids up to the base class

czue committed:
use correct new reference to location_id

czue committed:
convert one branch of aggregate stock status report to use elasticsearch

czue committed:
import cleanup [ci skip]

Commit:
Merge pull request #11899 from lwyszomi/228611-case

ILS: Wrap stock transactions saving into atomic transaction

millerdev committed:
Add 'web-user' kafka topic to topics.ALL

snopoke committed:
move details into body of email

snopoke committed:
Merge pull request #11617 from lwyszomi/133676-case

Allow to import/export gzipped json

Commit:
Merge pull request #12065 from dimagi/stock-status-scale

start converting commtrack reports to elasticsearch

orangejenny committed:
Bump checkbox over

Commit:
Merge pull request #12066 from dimagi/simplediff

simplify alembic diffs and fix email subject

Commit:
Merge pull request #11962 from dimagi/hqDefine-accounting

Hq define accounting

kaapstorm committed:
Merge pull request #12069 from dimagi/bump-checkbox

Bump checkbox over

millerdev committed:
Do not bake run.sh into docker image

With this change the image does not need to be rebuilt if run.sh or wait.sh are modified outside the container.

snopoke committed:
Merge pull request #11980 from dimagi/untangle-docker

Untangle docker

millerdev committed:
Create symlink for coverage report

dannyroberts committed:
Merge pull request #12061 from dimagi/new-account-km

adding km tracking for create account button

orangejenny committed:
vellum:7b6c26b3786f5e15dba7722114dc427c6ad185ac

Commit:
Merge pull request #12070 from dimagi/lighter-docker-image

Do not bake run.sh into docker image

benrudolph committed:
update staging.yaml

[ci skip]

benrudolph committed:
update staging.yaml

[ci skip]

orangejenny committed:
Merge pull request #12072 from dimagi/fix-coverage-report

Create symlink for coverage report

Commit:
Update submodules

rickypchen committed:
added tests for project_health

rickypchen committed:
Merge branch 'ppdreportfilter' of https://github.com/dimagi/commcare-hq into ppdreportfilter

dimagimon commented on corehq/form_processor/tests/test_case_dbaccessor.py:
```diff
@@ -102,43 +102,17 @@ def test_get_indices(self):
-
         referenced_case_ids = [uuid.uuid4().hex, uuid.uuid4().hex]
-        expected_case_ids = [_create_case_with_index(case_id) for case_id in referenced_case_ids]
+        _create_case_with_index(referenced_case_ids[0], case_is_deleted=True)  # case shouldn't be included in results
```
E501 line too long (118 > 115 characters)

dimagimon commented on corehq/apps/reports/tests/test_project_health.py:
```diff
@@ -0,0 +1,139 @@
+    now = datetime.datetime.utcnow()
+    year, month = add_months(now.year, now.month, 1)
+    month_as_date = datetime.date(year, month, 1)
+    prev_month_as_date = datetime.date(year, month-1, 1)
```
E226 missing whitespace around arithmetic operator

rickypchen commented:
https://github.com/dimagi/commcare-hq/pull/12100
follow-up PR
